### PR TITLE
fix: most immutable setting situtations

### PR DESCRIPTION
**Pull request by ilan-gold**
This is a PR to help push things along a bit here.  There are two commits here:

1. Bring back `backed` mode in its former state
2. Update the logic in initialization so that (a) `Raw` is initialized with an `AnnData` object that has a valid `layers` (hence the change of where the `Raw` declaration happens) and ~~(b) better handle the interaction between `layers` and `X`~~
3. EDIT: various small sharp edges

~~There are still quite a few errors but at least they fall into only a few categories I can address a bit more clearly now, I think.~~ Most of the remaining issues have to do with setting on a view or similar now, which is a bit fraught because setting on `X` is special-cased

One of the issues with the current logic is that `layers` now may or may not contain `None` but in `backed` mode, it never does even if `X` is not `None`, which is a problem for testing but I think also a bit of a rough edge.  I'm going to play around with holding a `SparseDataset` class in memory at `None` and seeing how that goes.

- [ ] Closes #
- [ ] Tests added
- [ ] Release note added (or unnecessary)

### CODE REVIEW

File: src/anndata/_core/anndata.py [hunk 1]
```diff
@@ -448,19 +451,24 @@ def _init_as_actual(  # noqa: PLR0912, PLR0913, PLR0915
...
-            self._raw = Raw(self, raw.X, raw.var, raw.varm)
 
         # clean up old formats
         self._clean_up_old_format(uns)
 
         # layers
+        if X is not None:
+            if layers is None or not bool(layers):
+                layers = {None: X}
+            elif bool(layers):
```
Comment (ilan-gold) [inline src/anndata/_core/anndata.py]: I think this can be `else`?

**Comment by ilan-gold** [comment]
>  I'm going to play around with holding a SparseDataset class in memory at None and seeing how that goes.

So the issue here is apparently doing anything after the object is a view and the file is closed.  You can't just reopen it so something like 

```python
backed_v = backed_adata[obs_idx, var_idx]
backed_v.to_memory() # closes file handle
backed_v.write_h5ad(final_pth)
```

doesn't work because `backed_v` has a closed file handle and thus can no longer read data.  We could change the behavior of `to_memory` to not close the file, but that would be a breaking change, and I'm not confident enough to say it is even a good idea.

File: src/anndata/_core/anndata.py (hunk 1 above)
Comment (flying-sheep) [reply src/anndata/_core/anndata.py]: yeah, if `None` and `False` mean the same, just don’t distinguish between them, they’re both falsy. Please read the last sentence of the first paragraph here: https://docs.python.org/3/reference/expressions.html#boolean-operations

Your logic here is:

```py
if layers:
    layers[None] = X
else:
    layers = {None: X}
```

is mutating the `layers` parameter OK?

File: src/anndata/_core/anndata.py [hunk 2]
```diff
@@ -537,15 +544,60 @@ def shape(self) -> tuple[int, int]:
...
+        else:
+            oidx, vidx = self._oidx, self._vidx
+        if (
+            np.isscalar(value)
+            or (hasattr(value, "shape") and (self.shape == value.shape))
+            or (self.n_vars == 1 and self.n_obs == len(value))
+            or (self.n_obs == 1 and self.n_vars == len(value))
+        ):
+            if self.is_view:
+                X = self.file["X"]
```
Comment (flying-sheep) [inline src/anndata/_core/anndata.py:587]: this is called even when `not self.isbacked` …

File: src/anndata/_core/anndata.py (hunk 1 above)
Comment (ilan-gold) [reply src/anndata/_core/anndata.py]: Hmm good point, wouldn't want to touch that.  Let me try to work around that

File: src/anndata/_core/anndata.py (hunk 2 above)
Comment (ilan-gold) [reply src/anndata/_core/anndata.py:587]: Yup, looking now

File: src/anndata/tests/helpers.py [hunk 3]
```diff
@@ -848,9 +848,17 @@ def fmt_name(x):
         "varp",
         "raw",
     ]:
+        a_elem, b_elem = getattr(a, attr), getattr(b, attr)
+        # TODO: This is helpful in backed mode where `X is not None` but `None not in layers`.
+        # Does this filter make sense in general? Is there a case where we explicitly want to check `None in layers`?
```
Comment (flying-sheep) [inline src/anndata/tests/helpers.py:853]: Yeah, backed mode complicates things.

My plan for the PR was that these two invariants should hold:
- `adata.X is adata.layers.get(None)`
- `(adata.X is None) is (None not in adata.layers)`

I get why you’re going the route of making `X`’s getter/setter more complex instead of `layers`, but I think my initial idea is more obvious from an API standpoint, don’t you think?

Comment (ilan-gold) [reply src/anndata/tests/helpers.py:853]: Totally agree, and I absolutely would go that way, but to do that we would need to deprecate parts of our current `X` functionality.  Things like
```python
import anndata as ad, numpy as np
adata = ad.AnnData(np.random.rand(3, 10), layers = { "a_layer":np.random.rand(3, 10) })
# updates adata.X
adata[:2].X = 1
# does not update adata.layers["a_layer"]
adata[:2].layers["a_layer"] = np.ones((2, 10))
```
actually do two different things so if were to route the former through the later, we would be dropping functionality that the former offers.  That's the same reason `adata.X is adata.layers.get(None)` doesn't hold; if it were to hold, we would be dropping functionality from what I can tell (in this case, `backed` mode more or less, where we never hold an array-like reference to the object, instead reopening the file on access).  Like I mentioned [here](https://github.com/scverse/anndata/pull/2095#issuecomment-3228407172), I would love to be able to remedy this, but we would be breaking even more currently standardized behavior then.

Maybe I'm missing what you're saying, but from my perspective, that was most of the idea with this PR - to bring back this functionality (largely).

Comment (ilan-gold) [reply src/anndata/tests/helpers.py:853]: In summation

1. backed mode currently has various edge cases, like never holding anything other than a reference to a file handle for `X`
2. Setting on `adata.X` has fundamentally different semantics than other elems

If we broke both of these, then I could see your invariants holding / complexity being removed.

Comment (flying-sheep) [reply src/anndata/tests/helpers.py:853]: yeah, that’s what I meant with 

> I get why you’re going the route of making `X`’s getter/setter more complex instead of `layers`

→ in order to make it all work transparently, we’d have to modify the `Layers*` classes more heavily to make all of this work the same.

but anyway, your approach here is “in backed mode, `.layers[None]` behaves as if `.X` was `None`”?

File: tests/test_views.py [hunk 4]
```diff
@@ -496,8 +496,15 @@ def test_view_delattr(attr, subset_func):
...
-    assert_equal(getattr(subset, attr), getattr(empty, attr))
+    # Should now have same value as default, except for `layers`, which still has the `None` key for `subset`
+    if attr == "layers":
+        assert_equal(
+            {k: v for k, v in getattr(subset, attr).items() if k is not None},
+            getattr(empty, attr),
+        )
+    else:
+        assert_equal(getattr(subset, attr), getattr(empty, attr))
+
```
Comment (flying-sheep) [inline tests/test_views.py:507]: I wonder if we should make `Layers` with only `None` inside falsey.

That way `del ad.layers; assert not layers` would work.
